### PR TITLE
Login cleanup scope reduction, restore original session if failed with switch

### PR DESCRIPTION
**Pull request by pzduniak**
Proof of concept, most likely will need a test for some of these flows that trigger it. Probably should fix all of these switch issues.

### CODE REVIEW

File: go/engine/login_provision.go [hunk 1]
```diff
@@ -1192,7 +1192,7 @@ func (e *loginProvision) cleanup(m libkb.MetaContext) {
 
 	// the best way to cleanup is to logout...
 	m.Debug("an error occurred during provisioning, logging out")
-	m.G().Logout(m.Ctx())
+	m.G().LogoutWithSecretKill(m, e.arg.User.GetNormalizedName(), true)
```
Comment (maxtaco) [inline go/engine/login_provision.go]: just delete this

File: go/engine/login_provision.go [hunk 2]
```diff
@@ -1192,7 +1192,7 @@ func (e *loginProvision) cleanup(m libkb.MetaContext) {
 
 	// the best way to cleanup is to logout...
 	m.Debug("an error occurred during provisioning, logging out")
-	m.G().Logout(m.Ctx())
```
Comment (maxtaco) [inline go/engine/login_provision.go]: there are 3 more of these to delete

File: go/libkb/globals.go [hunk 3]
```diff
@@ -307,11 +309,10 @@ func (g *GlobalContext) logoutSecretStore(mctx MetaContext, username NormalizedU
 	delete(g.switchedUsers, username)
 }
 
-func (g *GlobalContext) LogoutWithSecretKill(mctx MetaContext, killSecrets bool) (err error) {
+func (g *GlobalContext) LogoutWithSecretKill(mctx MetaContext, username NormalizedUsername, killSecrets bool) (err error) {
```
Comment (maxtaco) [inline go/libkb/globals.go]: Rename to `LogoutUsernameWithSecretKill`

File: go/engine/login.go [hunk 4]
```diff
@@ -125,7 +134,7 @@ func (e *Login) Run(m libkb.MetaContext) (err error) {
 	// clear out any existing session:
 	m.Debug("clearing any existing login session with Logout before loading user for login")
 	// If the doUserSwitch flag is specified, we don't want to kill the existing session
-	m.G().LogoutWithSecretKill(m, !e.doUserSwitch)
+	m.G().LogoutWithSecretKill(m, libkb.NewNormalizedUsername(e.username), !e.doUserSwitch)
```
Comment (maxtaco) [inline go/engine/login.go]: this is the wrong user to logout, you want the current username.

File: go/libkb/globals.go (hunk 3 above)
Comment (maxtaco) [reply go/libkb/globals.go]: Maybe also have a fuction `LogoutCurrentUserWithSecretKill`

Comment (maxtaco) [reply go/libkb/globals.go]: That way you don't have to change `Logout` or `ClearStateForSwitchUsers`

File: go/engine/login_provision.go (hunk 2 above)
Comment (pzduniak) [reply go/engine/login_provision.go]: Only 1 more with _fatal consequences_, the remaining calls are actually somewhat required (and hard to rework) but they are fixed with the restoreSession fn.

**Pull request merged**